### ISSUE TITLE
Conditionally export build directory to user package registry

### ISSUE THREAD

**Issue by mbeutel** [robotology/ycm#262]
`install_basic_package_files()` has great support for generating build-directory config files but doesn't currently export the build directory to the user package registry with [`export(PACKAGE)`](https://cmake.org/cmake/help/latest/command/export.html).

CMake 3.15 will change the behavior of `export(PACKAGE)` by introducing a new variable `CMAKE_EXPORT_PACKAGE_REGISTRY` and having `export(PACKAGE)` do nothing unless this variable is set to `ON` (cf. https://cmake.org/cmake/help/latest/policy/CMP0090.html).

I think we could nicely play along by adding a conditional `export(PACKAGE)` call like this:

```cmake

# option() honors normal variables since CMake 3.13 (cf. https://cmake.org/cmake/help/git-stage/policy/CMP0077.html).
if(POLICY CMP0077)
  cmake_policy(SET CMP0077 NEW)
endif()

option(CMAKE_EXPORT_PACKAGE_REGISTRY "Export build directory (enables external use without install)" OFF)


function(INSTALL_BASIC_PACKAGE_FILES _Name)
  ...
  # Export build directory if CMAKE_EXPORT_PACKAGE_REGISTRY is set.
  # This anticipates the introduction of the `CMAKE_EXPORT_PACKAGE_REGISTRY` variable in CMake 3.15 (cf. https://cmake.org/cmake/help/latest/policy/CMP0090.html).
  if(CMAKE_EXPORT_PACKAGE_REGISTRY)
    export(PACKAGE ${_Name})
  endif()
  ...
endfunction()
```

I can submit a pull request if you are interested.

**Comment by drdanz** [comment]
Sounds good to me, I'd just change
```diff
-if(CMAKE_EXPORT_PACKAGE_REGISTRY)
+if(CMAKE_EXPORT_PACKAGE_REGISTRY OR NOT CMAKE_VERSION VERSION_LESS 3.15)
```
so that it will be easier to spot and remove it when we will stop supporting CMake < 3.15

**Comment by traversaro** [comment]
I do not think we can avoid to call `export(PACKAGE ${_Name})` for CMake >= 3.15, as even if CMP0090 is set to TRUE and `CMAKE_EXPORT_PACKAGE_REGISTRY` is set to `ON`, `export(PACKAGE ${_Name})` needs to be explicitly called to populate the user package registry.

**Comment by drdanz** [comment]
The proposed change above is supposed to *always* call `export(PACKAGE ${_Name})` for CMake >= 3.15, or to call it depending on `CMAKE_EXPORT_PACKAGE_REGISTRY` on older releases.
Sorry if the `NOT VERSION_LESS` is confusing, but `VERSION_GREATER_EQUAL` was added in CMake 3.7 (if I remeber correctly), and until we make a decision on #261 we cannot use it...

**Comment by traversaro** [comment]
> The proposed change above is supposed to *always* call `export(PACKAGE ${_Name})` for CMake >= 3.15, or to call it depending on `CMAKE_EXPORT_PACKAGE_REGISTRY` on older releases.

Ah ok, I understood that you wanted the call to `export(..)`, my bad.

**Comment by drdanz** [comment]
Fixed by @mbeutel in #269.

**Comment by mbeutel** [comment]
> Sounds good to me, I'd just change
> 
> ```diff
> -if(CMAKE_EXPORT_PACKAGE_REGISTRY)
> +if(CMAKE_EXPORT_PACKAGE_REGISTRY OR NOT CMAKE_VERSION VERSION_LESS 3.15)
> ```
> 
> so that it will be easier to spot and remove it when we will stop supporting CMake < 3.15

Unfortunately that change made behavior inconsistent. The new behavior of `export(PACKAGE)` is implemented in CMake 3.15 and later, but it isn't *active* unless the consuming project also sets the policy version to 3.15 or later (which e.g. `cmake_minimum_required(VERSION 3.15)` implicitly does).

As an example, when using CMake 3.15 or later, the following program

```cmake
cmake_minimum_required(VERSION 3.14)
include(InstallBasicPackageFiles)
install_basic_package_files(foo
    COMPATIBILITY SameMajorVersion)
```

will register the build directory in the user package registry even if this wasn't requested.

I'll submit a PR to fix this by unconditionally checking `CMAKE_EXPORT_PACKAGE_REGISTRY` in `install_basic_package_files()`.
